presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -367,6 +376,24 @@ public void enforceMemoryLimits()
     }
 
     /**
+     * Enforce timeout at the query level
+     */
+    public void enforceQueryMaxRuntimeLimits()
+    {
+        Iterable<QueryExecution> runningQueries = queries.values().stream()
+                .filter(query -> query.getQueryInfo().getState() == RUNNING)
+                .collect(toImmutableList());
+
+        for (QueryExecution runningQuery : runningQueries) {
+            Duration queryMaxRuntime = SystemSessionProperties.getQueryMaxRuntime(runningQuery.getSession());
+            DateTime executionStartTime = runningQuery.getQueryInfo().getQueryStats().getExecutionStartTime();
```

Comment:
Will update the process logic accordingly after @haozhun  and @dain finish your discusstion.